Yatima/Typechecker/Datatypes.lean

```diff
@@ -35,12 +35,40 @@ inductive LiteralProp where
   | natLt (v1 v2 : Nat) (h : v1 < v2)
   | natNLt (v1 v2 : Nat) (h : ¬ v1 < v2)
 
+/--
+  The type info is a simplified form of the value's type, with only relevant
+  information for conversion checking, in order to get proof irrelevance and equality
+  of unit-like values.
+
+  - `unit` tells us that the expression's type is unit-like
+  - `proof` tells us that the expression's type is a proposition (belong to `Prop`)
+  - `prop` tells us that the expression's type is `Prop` itself
+-/
+inductive TypeInfo
+  | unit  : TypeInfo
+  | proof : TypeInfo
+  | prop  : TypeInfo
+  | none  : TypeInfo
+  deriving BEq, Inhabited
+
+/-- Representation of expressions for evaluation and transpilation -/
+inductive TypedExpr
+  | var   : TypeInfo → Name → Nat → TypedExpr
+  | sort  : TypeInfo → Univ → TypedExpr
+  | const : TypeInfo → Name → ConstIdx → List Univ → TypedExpr
+  | app   : TypeInfo → TypedExpr → TypedExpr → TypedExpr
+  | lam   : TypeInfo → Name → BinderInfo → TypedExpr → TypedExpr → TypedExpr
+  | pi    : TypeInfo → Name → BinderInfo → TypedExpr → TypedExpr → TypedExpr
+  | letE  : TypeInfo → Name → TypedExpr → TypedExpr → TypedExpr → TypedExpr
+  | lit   : TypeInfo → Literal → TypedExpr
+  | proj  : TypeInfo → ConstIdx → Nat → TypedExpr → TypedExpr
+  deriving BEq, Inhabited
```

Comment:
Would it be simpler to define this as:
```
inductive TypedExpr
  | mk : TypeInfo → Expr → TypedExpr
```
like you've done with `TypedValue` below?

Reply:
No, since the recursive instances must also contain `TypeInfo`. Notice that `Value` does not contain any direct recursive instance, only through `SusValue` which also contain `TypeInfo`. You could however, do something like

```
mutual
inductive TypedExpr
  | mk : TypeInfo → TypedExprInner → TypedExpr

inductive TypedExprInner
  | var   : Name → Nat → TypedExprInner
  | sort  : Univ → TypedExprInner
  | const : Name → ConstIdx → List Univ → TypedExprInner
  | app   : TypedExpr → TypedExpr → TypedExprInner
  | lam   : Name → BinderInfo → TypedExpr → TypedExpr → TypedExprInner
  | pi    : Name → BinderInfo → TypedExpr → TypedExpr → TypedExprInner
  | letE  : Name → TypedExpr → TypedExpr → TypedExpr → TypedExprInner
  | lit   : Literal → TypedExprInner
  | proj  : ConstIdx → Nat → TypedExpr → TypedExprInner
end
```

which is perhaps a better design, since you can extract the info without inspecting each case. Another possibility is parametrizing `Expr` as `ExprF` (the expression functor), defining `Expr = ExprF Expr` and `TypedExpr = ExprF (TypeInfo × TypedExpr)` (though since Lean is isorecursive, this requires some indirection through yet another inductive). But this makes `TypedExpr` definition inflexible, and in particular, I wouldn't be able to add `ConstIdx` to `TypedExpr.proj`